File: client/lib/src/models/message_record.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'message_record.freezed.dart';
part 'message_record.g.dart';

enum MessageContentType {
  @JsonValue('text')
  text,
  @JsonValue('image')
  image,
  @JsonValue('audio')
  audio,
  @JsonValue('video')
  video,
}

@freezed
abstract class MessageRecord with _$MessageRecord {
  const factory MessageRecord({
    required String id,
    @JsonKey(name: 'session_id') required String sessionId,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'character_id') required String characterId,
    required String role, // "user" | "ai" | "system"
    required String content,
    @JsonKey(name: 'content_type')
    @Default(MessageContentType.text)
    MessageContentType contentType,
    @JsonKey(name: 'action_type') String? actionType,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    Map<String, dynamic>? metadata,
  }) = _MessageRecord;

  factory MessageRecord.fromJson(Map<String, dynamic> json) =>
      _$MessageRecordFromJson(json);
}

File: app/schemas/message_record.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, Dict, Any
import uuid

from app.schemas.common import MessageContentMixin, MessageContentType

class MessageRecord(MessageContentMixin):
    id: str = Field(default_factory=lambda: str(uuid.uuid7()))
    session_id: str
    user_id: str
    character_id: uuid.UUID
    role: str  # "user" | "ai" | "system"
    # content, content_type, metadata inherited from MessageContentMixin
    action_type: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
